fixed expander bug, builtin optimisation

robertrinh opened the pull request:
1. fixed issue where echo $USER no quotes would segfault
2. updated export regarding checks
3. exit builtin optimalised
4. minor error_message fixes
5. minor free's
6. sources/executor/execute_commands/execute_piped_command.c check for main builtin needs to be deleted (forgot :))

QBeukelman commented on sources/builtins/export.c:
```diff
@@ -128,10 +130,14 @@ int		export(t_cmd *cmd, t_shell *shell)
-			add_arg_to_env(shell, cmd->args[i]);
-		else
-			show_error_message(E_EXPORT, C_RED, cmd->args[i], X_EXPORT);
+			if (is_valid_export_arg(cmd->args[i]))
```
`is_valid_export_arg()` should handle all checks. `count_delimiters()` should not be here.

robertrinh replied:
fair point. Issue was that export without the `export_delimiter` would segfault. Bash behavior would be this:

export a -> does nothing
export a in our minishell -> segfaults.

The idea was that the function should `return(0)` when no delimiter is found (and not an error!), essentially not doing anything like bash. Maybe you know a better approach? :O

robertrinh replied:
edit: fixed

robertrinh replied:
On second note: I put the only characters check in `is_valid_export_key`. As bash doesn't necessarily do anything with it, we can just argue that we handle that as an error instead. holds up the same result, not exporting anything in the env!

QBeukelman commented on sources/executor/execute_commands/execute_piped_command.c:
```diff
@@ -22,9 +22,13 @@ t_validation	execute_piped_command(t_shell *shell, t_cmd *cmd)
 	cmd_path = cmd->cmd_path;
 	formatted_cmd = cmd->formatted_cmd;
 
+	if (is_builtin(shell->builtin_main, cmd, B_NUM_MAIN))
```
Are these commands not meant to execute in the main thread?

robertrinh replied:
Correct, I have mentioned that in the comment -> bulletpoint 6

QBeukelman commented on sources/executor/execute_commands/execute_wait.c:
```diff
@@ -6,25 +6,17 @@
 
 #include "../../../includes/minishell.h"
 
-static int	handle_signal(int signal)
```
Keep this function. I am still working on the exit codes here.

robertrinh replied:
my bad! My opinion is that the signals are handled correctly in the childprocess and I dont believe it's neccesary to handle the signals at waitpid function. will revert it

QBeukelman commented on sources/executor/execute_commands/execute_wait.c:
```diff
@@ -41,12 +33,10 @@ int	wait_for_last_cmd(int child_count, int last_pid)
 	}
 	if (WIFEXITED(last_status))
 		return (WEXITSTATUS(last_status));
-	else if (WIFSIGNALED(last_status))
```
Keep this

robertrinh replied:
see previous comment

QBeukelman commented on sources/expander/expander.c:
```diff
@@ -37,8 +37,9 @@ static char	*expand_arg(char **env, char *arg, size_t i)
 	{
-		free_env_values(env_utils->key, env_utils->value, env_utils);
-		return (arg = env_utils->value);
+		arg = env_utils->value;
```
All values need to be passed to`free_env_values()` function.

robertrinh replied:
Issue is that freeing the values in this specific statement that the arg value would be lost, having no value.

[minishell]: echo $USER
� 
[minishell]: echo "$USER"
qtrinh 

you can try it out. not passing the env_utils->value to free_env_values retains the expanded env value and therefore, works correctly.

QBeukelman reviewed (changes requested):
* Incorrect branch name: "bugfix/#102-Branch-name"
* Incorrect commit message: "bugfix/#102 Commit message"

robertrinh commented:
1. The stuff in the prior comments are updated accordingly except expander (will discuss!)
2. fixed issue #100 where export stops giving error after multiple inputs
3. fixed issue mentioned in #85 that the minishell outputted previous command when pressing `SPACEBAR`
4. Started combatting memory issues, values initialised
5. Fixed few memory issues consisting `''Conditional jump or move depends on uninitialised value(s)''`
6. fixed issue where original_fd kept incrementing after a command
7. #85 in progress.
8. removed fd_err in the cmd struct, as we have not used it at all.
